resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RateLimiterConfig.java

```diff
@@ -161,6 +173,21 @@ public Builder writableStackTraceEnabled(boolean writableStackTraceEnabled) {
             return this;
         }
 
+        /**
+         * Allows you to check the result of a call decorated by this rate limiter and make a decision
+         * should we drain all the permissions left it the current period. Useful in situations when
+         * despite using a RateLimiter the underlining called service will say that you passed the maximum number of calls for a given period.
+         *
+         * @param drainPermissionsOnResult your function should return true when the permissions drain
+         *                                 should happen
+         * @return the RateLimiterConfig.Builder
+         * @see RateLimiter#drainPermissions()
+         */
+        public Builder drainPermissionsOnResult(Function<CallsResult, Boolean> drainPermissionsOnResult) {
```

Comment:
Please use `Either<Throwable, Object>` instead of `CallsResult`

Reply:
are you sure? CallsResult has a nicer API which enables cleaner configuration 

please take a look at my unit test:

resilience4j-ratelimiter/src/test/java/io/github/resilience4j/ratelimiter/RateLimiterWithConditionalDrainIntegrationTest.java

the configuration part would not look so simple if I used `Either<Throwable, Object>`

Reply:
Yes, I would prefer to use Either for now.
We are working on a 2.0 branch which uses Java 11 and has no Vavr dependency anymore. We will replace Either by our own implementation at many places.  This would be the best time to introduce a simpler API.

Reply:
I've tried refactoring the code to `Either` but `Either` cannot handle a successful call return of a function that returns `void`

this is needed in situations like

```
    static CheckedRunnable decorateCheckedRunnable(RateLimiter rateLimiter, int permits,
        CheckedRunnable runnable) {

        return () -> {
            waitForPermission(rateLimiter, permits);
            try {
                runnable.run();
                drainIfNeeded(rateLimiter, CallsResult.success());
            } catch (Exception exception) {
                drainIfNeeded(rateLimiter, CallsResult.failure(exception));
                throw exception;
            }
        };
    }
```

what should I use here? `Either.right(null)` ?

and then in the application code which check for a certain condition should I do this:

```
if (result.isRigth() && result.get() != null) {
```

?

besides that the API started looking ugly with template mayhem such as:

```
private final Function<Either<? extends Throwable, Object>, Boolean> drainPermissionsOnResult;
```

I do not think this is the right approach.

Furthermore where should I put helper methods like `isSuccessfulAndReturned`:

```
        RateLimiter limiter = RateLimiter.of("someLimiter", RateLimiterConfig.custom()
            .limitForPeriod(5)
            .limitRefreshPeriod(Duration.ofHours(1))
            .drainPermissionsOnResult(
                callsResult -> callsResult.isSuccessfulAndReturned(
                    ResponseWithOverloadIndication.class,
                    response -> response.isPotentialOverload()))
            .build());
```

?

Reply:
We can add a `ResultUtil` class.

```
ResultUtil.isSuccessfulAndReturned(eitherResult, responseWithOverloadIndication.class,  response -> response.isPotentialOverload())
```

As I said, when we remove Vavr from Resilience4j 2.0, we will have a look at the whole code base and might introduce a call like you suggest. But then we will use it everywhere and consistently. 
In v1.x, we are using `Either` when a user needs to check if the call was a sucess or failure.

In v1.x anew `ResultUtil` helper class could also be utilized in `IntervalBiFunction`.

If you don't like it, as an alternative we could introduce:

`RateLimiterConfig.drainPermissionsOnResult` and `RateLimiterConfig.drainPermissionsOnException` comparable to `RetryConfig.retryOnException` and `RetryConfig.retryOnResult`.

Reply:
removed `CallsResult` introduced `Either`